icuSources/CMakeLists.txt

```diff
@@ -0,0 +1,519 @@
+    i18n/numrange_fluent.cpp
+    i18n/numrange_impl.cpp
+    i18n/numsys.cpp
+    i18n/olsontz.cpp
+    i18n/persncal.cpp
+    i18n/pluralranges.cpp
+    i18n/plurfmt.cpp
+    i18n/plurrule.cpp
+    i18n/quant.cpp
+    i18n/quantityformatter.cpp
+    i18n/rbnf.cpp
+    i18n/rbt.cpp
+    i18n/rbt_data.cpp
+    i18n/rbt_pars.cpp
+    i18n/rbt_rule.cpp
+    i18n/rbt_set.cpp
+    i18n/rbtz.cpp
+    i18n/regexcmp.cpp
+    i18n/regeximp.cpp
+    i18n/regexst.cpp
+    i18n/regextxt.cpp
+    i18n/region.cpp
+    i18n/reldatefmt.cpp
+    i18n/reldtfmt.cpp
+    i18n/rematch.cpp
+    i18n/remtrans.cpp
+    i18n/repattrn.cpp
+    i18n/rulebasedcollator.cpp
+    i18n/scientificnumberformatter.cpp
+    i18n/scriptset.cpp
+    i18n/search.cpp
+    i18n/selfmt.cpp
+    i18n/sharedbreakiterator.cpp
+    i18n/simpletz.cpp
+    i18n/smpdtfmt.cpp
+    i18n/smpdtfst.cpp
+    i18n/sortkey.cpp
+    i18n/standardplural.cpp
+    i18n/string_segment.cpp
+    i18n/strmatch.cpp
+    i18n/strrepl.cpp
+    i18n/stsearch.cpp
+    i18n/taiwncal.cpp
+    i18n/timezone.cpp
+    i18n/titletrn.cpp
+    i18n/tmunit.cpp
+    i18n/tmutamt.cpp
+    i18n/tmutfmt.cpp
+    i18n/tolowtrn.cpp
+    i18n/toupptrn.cpp
+    i18n/translit.cpp
+    i18n/transreg.cpp
+    i18n/tridpars.cpp
+    i18n/tzfmt.cpp
+    i18n/tzgnames.cpp
+    i18n/tznames.cpp
+    i18n/tznames_impl.cpp
+    i18n/tzrule.cpp
+    i18n/tztrans.cpp
+    i18n/uameasureformat.cpp
+    i18n/uatimeunitformat.cpp
+    i18n/ucal.cpp
+    i18n/ucln_in.cpp
+    i18n/ucol.cpp
+    i18n/ucol_res.cpp
+    i18n/ucol_sit.cpp
+    i18n/ucoleitr.cpp
+    i18n/ucsdet.cpp
+    i18n/udat.cpp
+    i18n/udateintervalformat.cpp
+    i18n/udatintv.cpp
+    i18n/udatpg.cpp
+    i18n/ufieldpositer.cpp
+    i18n/uitercollationiterator.cpp
+    i18n/ulistformatter.cpp
+    i18n/umsg.cpp
+    i18n/unesctrn.cpp
+    i18n/uni2name.cpp
+    i18n/units_complexconverter.cpp
+    i18n/units_converter.cpp
+    i18n/units_data.cpp
+    i18n/units_router.cpp
+    i18n/unum.cpp
+    i18n/unumsys.cpp
+    i18n/uplrule.cpp
+    i18n/upluralrules.cpp
+    i18n/uregex.cpp
+    i18n/uregexc.cpp
+    i18n/uregion.cpp
+    i18n/usearch.cpp
+    i18n/uspoof.cpp
+    i18n/uspoof_build.cpp
+    i18n/uspoof_conf.cpp
+    i18n/uspoof_impl.cpp
+    i18n/utf16collationiterator.cpp
+    i18n/utf8collationiterator.cpp
+    i18n/utmscale.cpp
+    i18n/utrans.cpp
+    i18n/vtzone.cpp
+    i18n/vzone.cpp
+    i18n/windtfmt.cpp
+    i18n/winnmfmt.cpp
+    i18n/wintzimpl.cpp
+    i18n/zonemeta.cpp
+    i18n/zrule.cpp
+    i18n/ztrans.cpp
+
+    # io sources
+    io/locbund.cpp
+    io/sprintf.cpp
+    io/sscanf.cpp
+    io/ucln_io.cpp
+    io/ufile.cpp
+    io/ufmt_cmn.cpp
+    io/uprintf.cpp
+    io/uprntf_p.cpp
+    io/uscanf.cpp
+    io/uscanf_p.cpp
+    io/ustdio.cpp
+    io/ustream.cpp)
+
+target_include_directories(_FoundationICU
+    PRIVATE
+        common/
+        i18n/
+        io/
+    PUBLIC
+        include/)
+
+target_link_libraries(_FoundationICU PRIVATE stdc++)
```

Comment:
This is an odd one since we have to guess what flavor of C++ folks are using since the Swift driver decided to not help us out and tell us what it's using (actually, I think part of the problem is that it doesn't know what flavor of C++ we're using either leaving it as an exercise to the user to pass all the right flags).
I know this isn't going to work for Android (uses libcxx) nor for Windows (which can use the Windows C++ runtime, but I think libcxx might also work there too so it could go either way). I think it should be customizable at distribution time since the person putting together the SDK will likely have the best chance of knowing which C++ runtime they're building the SDK with. At least until the Swift compiler learns how to specify what C++ runtime you're working with.

In theory, we could do something like this to provide reasonable defaults:
```cmake
if(Linux)
set(SwiftFoundation_CXX_RUNTIME_default stdc++)
elseif(Android)
set(SwiftFoundation_CXX_RUNTIME_default c++)
elseif(WIN32)
set(SwiftFoundation_CXX_RUNTIME_default WhateverWindowsUses)
endif()
set(SwiftFoundation_CXX_RUNTIME "${SwiftFoundation_CXX_RUNTIME_default}" CACHE STRING
    "C++ runtime that SwiftFoundation is built against")
```

@compnerd, do you have a better idea?

Reply:
Why not leave it to the user? If we set the linker language to C++, the C++ driver will be used. The user can add in `-stdlib=libc++` or `-stdlib=libstdc++` if they are using clang. If they are using cl, that will default to the MS STL, as would `clang-cl`.

Reply:
This is related to using c++ interop at link time. The one thing that we should be careful of is if that gets serialized into the module.

Reply:
I only added this because I was getting linking errors otherwise in downstream projects (Yams in this case):
```
FAILED: bin/YamsTestRunner Tests/CMakeFiles/YamsTestRunner.dir/LinuxMain.swift.o swift/YamsTestRunner.swiftmodule
: && /Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/bin/swiftc -output-file-map Tests/CMakeFiles/YamsTestRunner.dir/Debug/output-file-map.json -incremental -j 12 -emit-executable -o bin/YamsTestRunner -emit-dependencies  -module-cache-path "/Users/charleshu/Developer/FCF/SwiftStaging/build/buildbot_linux/swiftpm-linux-aarch64/aarch64-unknown-linux-gnu/yams/module-cache" -g  -module-cache-path "/Users/charleshu/Developer/FCF/SwiftStaging/build/buildbot_linux/swiftpm-linux-aarch64/aarch64-unknown-linux-gnu/yams/module-cache" -g -I /Users/charleshu/Developer/FCF/SwiftStaging/build/buildbot_linux/swiftpm-linux-aarch64/aarch64-unknown-linux-gnu/yams/swift /Users/charleshu/Developer/FCF/SwiftStaging/yams/Tests/LinuxMain.swift  -L /Users/charleshu/Developer/FCF/SwiftStaging/build/buildbot_linux/swiftpm-linux-aarch64/aarch64-unknown-linux-gnu/yams/lib  -L /Users/charleshu/Developer/FCF/SwiftStaging/build/buildbot_linux/swiftpm-linux-aarch64/aarch64-unknown-linux-gnu/yams/lib -Xlinker -rpath -Xlinker /Users/charleshu/Developer/FCF/SwiftStaging/build/buildbot_linux/swiftpm-linux-aarch64/aarch64-unknown-linux-gnu/yams/lib  lib/libYamsTests.so  -lXCTest  -lFoundation  -lXCTest  lib/libYams.so && :
error: link command failed with exit code 1 (use -v to see invocation)
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'vtable for __cxxabiv1::__vmi_class_type_info'
/usr/bin/ld.gold: the vtable symbol may be undefined because the class is missing its key function
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'std::condition_variable::wait(std::unique_lock<std::mutex>&)'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'std::condition_variable::notify_all()'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'std::condition_variable::~condition_variable()'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to '__dynamic_cast'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'std::__throw_system_error(int)'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'std::condition_variable::condition_variable()'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'operator new(unsigned long)'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'std::__once_callable'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'std::__once_call'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to '__once_proxy'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to '__cxa_guard_abort'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to '__cxa_guard_release'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to '__cxa_guard_acquire'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'operator new[](unsigned long)'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'operator delete[](void*)'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'std::__throw_length_error(char const*)'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_append(char const*, unsigned long)'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'operator delete(void*)'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::resize(unsigned long, char)'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'vtable for __cxxabiv1::__class_type_info'
/usr/bin/ld.gold: the vtable symbol may be undefined because the class is missing its key function
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to '__cxa_pure_virtual'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to '__gxx_personality_v0'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'vtable for __cxxabiv1::__si_class_type_info'
/usr/bin/ld.gold: the vtable symbol may be undefined because the class is missing its key function
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to 'std::terminate()'
/Users/charleshu/Developer/FCF/SwiftStaging/Swift/swift-nightly-install/usr/lib/swift/linux/libFoundationInternationalization.so: error: undefined reference to '__cxa_begin_catch'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```